src/lib/krb5/ccache/cc_file.c

```diff
@@ -1002,12 +1018,144 @@ fcc_store(krb5_context context, krb5_ccache id, krb5_creds *creds)
     return set_errmsg_filename(context, ret ? ret : ret2, data->filename);
 }
 
-/* Non-functional stub for removing a cred from the cache file. */
+/*
+ * Overwrite cred in the ccache file with an entry that should not match any
+ * reasonable search.  Deletion is not guaranteed.  This algorithm originally
+ * from Heimdal.
+ */
+static krb5_error_code
+cred_delete(krb5_context context, krb5_ccache cache, krb5_cc_cursor *cursor,
+            krb5_creds *cred)
+{
+    krb5_error_code ret;
+    krb5_fcc_cursor *fcursor = *cursor;
+    fcc_data *fdata = cache->data;
+    struct k5buf expected = EMPTY_K5BUF, overwrite = EMPTY_K5BUF;
+    int fd = -1;
+    uint8_t *on_disk = NULL;
+    ssize_t rwret, start_offset;
+
+    k5_buf_init_dynamic_zap(&expected);
+    k5_buf_init_dynamic_zap(&overwrite);
+
+    /* This is what the cred looks like in the file (we hope). */
+
+    k5_marshal_cred(&expected, fcursor->version, cred);
+    ret = k5_buf_status(&expected);
+    if (ret)
+        goto cleanup;
+
+    /* Mark the cred expired so that it will hopefully be skipped over by any
+     * future match checks. */
+    cred->times.endtime = 0;
+
+    /* For config creds, endtimes aren't checked, so change the realm.
+     * Heimdal uses X-RMED-CONF here, so we do too. */
+    if (cred->server != NULL && cred->server->realm.length > 0 &&
+        strcmp(cred->server->realm.data, "X-CACHECONF:") == 0)
+        strcpy(cred->server->realm.data, "X-RMED-CONF:");
+
+    k5_marshal_cred(&overwrite, fcursor->version, cred);
+    ret = k5_buf_status(&overwrite);
+    if (ret)
+        goto cleanup;
+
+    if (expected.len != overwrite.len) {
+        ret = KRB5_CC_FORMAT;
+        goto cleanup;
+    }
+
+    /* Get a non-O_APPEND handle to the raw file. */
+    fd = open(fdata->filename, O_RDWR);
+    if (fd == -1) {
+        ret = interpret_errno(context, errno);
+        goto cleanup;
+    }
+
+    ret = krb5_lock_file(context, fd, KRB5_LOCKMODE_EXCLUSIVE);
```

Comment:
I don't think we need to lock; the worst that happens is we race with another thread or process trying to remove the same credential, performing the same write operation.

Reply:
On further reflection, the situation here is muddier than I thought, because of the sanity check.  Two processes racing to delete the same cred will be trying to perform the same write, but one of them might not see the expected contents when it reads what was previously there.  This won't corrupt the ccache, but will produce a spurious KRB5_CC_FORMAT error.

However, locking here does not solve that problem, because the lock comes too late--we read in the previous cred entry before obtaining the lock, so another process could get in and remove the cred before our process obtains the write lock.

(Also, I'm surprised that obtaining a write lock here doesn't deadlock against the read lock obtained for iteration, on kernels recent enough to have OFD locking.)

Options I can see include:

1. Remove the sanity check, trusting that it wouldn't fail in practice anyway.  If we do this, I would like to pass in the start_offset from fcc_remove_cred() (which can do an ftell() on the file handle before each call to fcc_next_cred()) and sanity-check the difference in file offsets against the length of the new marhshalled cred.

2. Instead of comparing byte for byte, unmarshal the old byte values and check it... or don't check it, because if it unmarshals successfully, we at least skipped back to the beginning of a cred entry.

3. Arrange to obtain the write lock when fcc_remove_cred() starts iterating.  Not my preferred option.

Reply:
The file descriptor isn't kept locked between iteration calls, so there's no contention and therefore no deadlock.

Heimdal's approach is best-effort to the point that `cred_delete()` can't pass an error back up.  If we're not getting a lock from `fcc_remove_cred()`, it seems like the best option is to make any failure here graceful.  How would you feel about:

4. Make the `memcmp()` failure a non-failure - just return 0 there instead of `KRB5_CC_FORMAT`.

Reply:
I think that would be okay too.  If the byte comparison doesn't succeed, either another process removed the same cred before we got to it, or the cache was reinitialized.  Either way, the cred is no longer present in the cache.  (I hadn't thought about the possibility of the cache being reinitialized, or I wouldn't have listed option 1 and would have phrased option 2 differently.)

(Well, unless our marshalling functions aren't perfect reversals of each other.  If we had a good way to save the result of load_cred() while iterating that wouldn't be a concern.  But the marshalling functions should round-trip correctly.)